src/core/mac/sub_mac_wed.cpp

```diff
@@ -0,0 +1,91 @@
+/*
+ *  Copyright (c) 2024, The OpenThread Authors.
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions are met:
+ *  1. Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *  2. Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *  3. Neither the name of the copyright holder nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   This file implements the Wake-up End Device of the subset of IEEE 802.15.4 MAC primitives.
+ */
+
+#include "sub_mac.hpp"
+
+#if OPENTHREAD_CONFIG_WAKEUP_END_DEVICE_ENABLE
```

Comment:
This part of code is running on the Host side in RCP mode, right? It will prevent the Thread radio chip from entering listening mode at the precise time. Because the Spinel communication delay between the host and the RCP is not a fixed value (when the Host CPU is busy, the delay is large).  Could we make this part of code running on the RCP side?

Reply:
Isn't it already the case since we are adding the file to `RADIO_COMMON_SOURCES`?

Reply:
This file is also be added to `COMMON_SOURCES`. Then the `Mac` will directly call functions defined in this file on the Host side. This part of code won't run on the RCP side.

Reply:
What's your proposal then? If I remove it from `COMMON_SOURCES` non-RCP modes won't build.

Reply:
I think we can add a radio capability `OT_RADIO_CAPS_WED`, a platform function `otPlatRadioUpdateWakeupListening()` and a link raw API `otLinkRawUpdateWakeupListening()`. When the `SubMac` detects that the RCP supports `OT_RADIO_CAPS_WED`, the function `SubMac::UpdateWakeupListening()` calls the `otPlatRadioUpdateWakeupListening()`. The `otPlatRadioUpdateWakeupListening()` sends a new defined Spinel command to RCP, the receiver of the Spinel command calls the `otLinkRawUpdateWakeupListening()` and the `otLinkRawUpdateWakeupListening()` calls the `SubMac::UpdateWakeupListening()` on the RCP side.

It could be implemented in a separated commit.

Reply:
Okay, looks good, but yes, lets move it to a future task in order to unlock other contributions.